pennylane/beta/queuing/queuing.py

```diff
@@ -0,0 +1,248 @@
+    queuing functions queue external objects.
+
+    Queuable objects can take use of the following ``QueuingContext`` methods and properties:
+
+    * :attr:`~.recording`: determine whether a queuing context is actively recording operations
+    * :meth:`~.append`: append an object to the active queuing context
+    * :meth:`~.remove`: remove an object from the queuing context. If the object is queued multiple
+      times, only the first occurrence is removed.
+    * :meth:`~.update_info`: updates metadata attached to an object in the queue (not supported by
+      all queuing contexts).
+    * :meth:`~.get_info`: retrieves metadata attached to an object in the queue (not supported by
+      all queuing contexts).
+
+    Queuing context subclasses must define the following abstract methods:
+
+    * :meth:`~._append`: define an action to perform when an object append
+      request is made.
+
+    * :meth:`~._remove`: define an action to perform when an object removal request is made.
+
+    In addition, the optional methods :meth:`~._update_info` and ``get_info`` may also be
+    defined as required.
+
+    **Example**
+
+    To create a queuing context, simply subclass from ``QueuingContext`` and define
+    the required methods:
+
+    >>> class MyQueue(QueuingContext):
+    ...     def __init__(self):
+    ...         self.queue = []
+    ...     def _append(self, obj):
+    ...         self.queue.append(obj)
+    ...     def _remove(self, obj):
+    ...         self.queue.remove(obj)
+
+    Once defined, it can be used as a queuing context to track queuable objects
+    and queuing functions:
+
+    >>> with MyQueue() as q:
+    ...     QueuingContext.append("object")
+    >>> print(q.queue)
+    ['object']
+
+    Note that ``QueuingContext`` subclasses support nesting; objects are only queued to the
+    first surrounding queuing context:
+
+    >>> with MyQueue() as q1:
+    ...     with MyQueue() as q2:
+    ...         QueuingContext.append("first object")
+    ...     QueuingContext.append("second object")
+    ...
+    >>> print(q1.queue)
+    ['second object']
+    >>> print(q1.queue)
+    ['first object']
+
+    Finally, queuing contexts *themselves* can be queuable objects:
+
+    >>> class QueuableQueue(QueuingContext):
+    ...     def __init__(self):
+    ...         self.queue = []
+    ...         QueuingContext.append(self)
+    ...     def _append(self, obj):
+    ...         self.queue.append(obj)
+    ...     def _remove(self, obj):
+    ...         self.queue.remove(obj)
+
+    We can see that nested ``QueuableQueue`` objects are queued to their surrounding queuing
+    context:
+
+    >>> with QueuableQueue() as q1:
+    ...     with QueuableQueue() as q2:
+    ...         QueuingContext.append("first object")
+    ...     QueuingContext.append("second object")
+    >>> print(q1.queue)
+    [<__main__.QueuableQueue object at 0x7f94c432b6d0>, 'second object']
```

Comment:
Yep! The difference here is that the `QueuableQueue` now calls `QueuingContext.append(self)` on init.

So a bit of additional work for the developer, but I think it makes sense, and is a clean abstraction. It allows for a lot of flexibility; the developer can decide when the context queues itself. Whether this is on init, enter, or exit.

Reply:
I think I don't get it, what is the use case of nested queues? 

I am asking because a nested tape could be quite a beautiful representation for a "batch of circuits" that can be nicely dispatched to a device by the tape. From Josh's example,

``` python 
with QuantumTape() as tape:
    qml.BasisState(np.array([1, 1]), wires=[0, 'a'])

    with QuantumTape() as tape2:
        qml.Rot(0.543, 0.1, 0.4, wires=0)
        qml.Rot(0.3, 0.0, 0.0, wires=0)
        qml.Hadamard(wires=0)

    qml.CNOT(wires=[0, 'a'])
    qml.RY(0.2, wires='a')
    probs(wires=0), probs(wires='a')
```
could be interpreted as

``` python
>>> tape.operations
[BasisState(array([1, 1]), wires=[0, 'a']),
<QuantumTape: params=6, wires=<Wires = [0]>>,
CNOT(wires=[0, 'a']),
RY(0.2, wires=['a'])]
```
which expands to three "paths through the tape":

``` python
>>> tape.expand()
[BasisState(array([1, 1]), wires=[0, 'a']),
Rot(0.543, 0.1, 0.4, wires=[0]),
CNOT(wires=[0, 'a']),
RY(0.2, wires=['a'])]

[BasisState(array([1, 1]), wires=[0, 'a']),
Rot(0.3, 0.0, 0.0, wires=[0]),
CNOT(wires=[0, 'a']),
RY(0.2, wires=['a'])]

[BasisState(array([1, 1]), wires=[0, 'a']),
Hadamard(wires=0),
CNOT(wires=[0, 'a']),
RY(0.2, wires=['a'])]
```
I am not sure how the latter would be represented, since it is a list of queues? But these three circuits constructed from different parameters and/or gates can then be sent to a device by ``tape.execute(device)``, without changing any syntax.

We could provide user UI to nest their quantum functions this way (i.e. by comparing the quantum function queues and nesting only the parts that differ), and we could have the param shift rules just be "nesting rules".

Is that a crazy thought?

Reply:
The current use case of nested queues would be decompositions; for example, if a device doesn't support `Rot`, the `Rot` gate would be replaced with a queue containing the decomposition, and then expanded. So I think there is still a usecase to having all options in a nested tape simply join the parent tape.

What you suggest is cool however --- I think I have seen this in other libraries? For example, `tape.expand()` 'subsumes' nested tapes, whereas something like `tape.explode()` generate numerous tapes, by treating each operation in subtapes as a separate pathway.

I think we would still have to think of the optimum way of specifying which ops are different pathways, and which aren't

Reply:
Can't we just a ``BatchTape`` for those that should be interpreted as parallel code, and ``explode`` only looks at the ``BatchTape`` nesting?

Reply:
I like that!

Reply:
Nesting was never really considered as part of any "batching" feature, because the nested tapes all represent the *same* circuit, but at different levels of granularity. Batching would be a somewhat independent feature (e.g., using something like a list of tapes)

Reply:
Yes, but let's say we have a batch of 1 Million quantum circuits each executed by a one-shot measurement (which appears in our current ideas around making single-shot measurements diffable). Would you really want to represent this by 1 Mio tapes and send it to the backend? Or worse, by 1 Mio Qnodes? 

If parameters _and_ gate types/templates _and_ observables are batched, for example in a VQE-like stochastic gradient descent gridsearch problem, we quickly get large batches. Wouldn't it be great to represent them as a single circuit where only the part that differs is represented by a ``BatchTape``? 

                                  StatePrep()
                                  CNOT(wires=[0,3])
                                           |
                       ---------------------------------------
                      |                    |                          |
  RY(0.1, wires=0)   RX(0.1, wires=0)   RZ(0.1, wires=0)
                      |                    |                          |
                       ---------------------------------------
                                           |
                                    CNOT(wires=[2, 1])
                                           |
                       ---------------------------------------
                      |                    |                          |
  RX(0.1, wires=0)   RX(0.2, wires=0)   RX(0.3, wires=0)
                      |                    |                          |
                       ---------------------------------------
                                           |
            -----------------------------------------------------------------------------
           |                                           |                                                |                                      
  expval(PauliZ(), shots=1)   expval(PauliZ(), shots=4)   expval(PauliX(), shots=1)

This is an exponentially compact representation which could be unfolded at will by the device before sending it off. In fact, this idea would solve so many of my current problems around shots and paralellisation with a very gentle extension of the tape structure, I would like to try a prototype if no one minds...

Reply:
Another efficiency gain: compilers could act on the compact representation, instead of compiling 1Mio almost-copies of a circuit!